### PR TITLE
Mock server

### LINKED ISSUE #164: Create a MANY server mock crate
This can even be part of the `many` cli tool. To facilitate with development against MANY modules that don't exist yet.

### PR DESCRIPTION
**Pull request by Denommus**
## Description

A user can now use a configuration TOML file containing keys and values to assign mock responses to the CLI running in server mode.

This file can either be within the PWD where the server is running from and be called `mockfile.toml`, or it can be passed as the argument `--mockfile`.

## Related Issue

Fixes #164

## Testing

The tests are done via unit testing in the PR itself

## Checklist:

- [X] I have read and followed the CONTRIBUTING guidelines for this project.
- [X] I have added or updated tests and they pass.
- [X] I have added or updated documentation and it is accurate.
- [X] I have noted any breaking changes in this module or downstream modules.

### CODE REVIEW

File: src/many-mock/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+/// Reads and parses the mockfile provided by the mockfile_arg parameter, or from a default path
+pub fn parse_mockfile(
+    mockfile_arg: Option<&str>,
+) -> Result<MockEntries, Box<dyn std::error::Error>> {
+    let contents = if let Some(mockfile_arg) = mockfile_arg {
+        // A file was passed from the command line
+        std::fs::read_to_string(mockfile_arg)?
+    } else {
+        // If no file was passed from the command line, read the default path if it exists
```
Comment (fmorency) [inline src/many-mock/src/lib.rs:21]: Why not put the default value in the clap option?

**Review by fmorency — CHANGES_REQUESTED** [review]
Thanks, Yuri! This is a good first iteration.

I would like to see a real mock file example/test(?), used to interact with real MANY clients!

File: src/many-mock/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        let fallback_path = std::path::Path::new("mockfile.toml");
+        if fallback_path.exists() {
+            std::fs::read_to_string(fallback_path)?
+        } else {
+            // If there's no file passed either from the command line or as a fallback path, just use an empty content
+            "".to_string()
```
[the remark this answers is not in the document]
Comment (Denommus) [reply src/many-mock/src/lib.rs:27]: I'm assuming one that doesn't pass a mockfile will not want any mock endpoints, that's why I decided for an empty content.

File: src/many-mock/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+pub fn fill_placeholders(request: &RequestMessage, response: String) -> String {
+    let map = load_request(request);
+    map.iter().fold(response, |acc, (key, value)| {
+        let re = regex::Regex::new(key).unwrap();
+        re.replace_all(&acc, value).to_string()
+    })
+}
+
+#[cfg(test)]
+mod tests {
```
[the remark this answers is not in the document]
Comment (Denommus) [reply src/many-mock/src/lib.rs:74]: I agree, but in this case, I'm using tests as feature documentation, not to really enforce anything.

**Comment by Denommus** [comment]
@fmorency I believe your comments were addressed

**Comment by fmorency** [comment]
As discussed on Slack,

> I would like to see a real mock file example/test(?), used to interact with real MANY clients!

**Review by fmorency — CHANGES_REQUESTED** [review]
I like where this is going! Thanks, @Denommus!

I would like you to add a mock test file for the product team which implements every `many-ledger` endpoint. This will allow them to test their Gwen stuff. You could also add a BATs test to test this stuff or any other way you can think of?

I would also like you to add your `rspec` tests to the CI. Right now, those tests are not run, and we don't see any impact on the code coverage.

I'll let @hansl pitch in.

File: src/many-server/src/server.rs [hunk 4]
```diff
@@ -326,11 +329,28 @@ impl LowLevelManyRequestHandler for Arc<Mutex<ManyServer>> {
                     LowLevelManyRequestHandler::execute(fb.as_ref(), envelope).await
                 }
                 (None, None) => {
-                    let response = ResponseMessage::error(
-                        &cose_id.identity,
-                        id,
-                        ManyError::could_not_route_message(),
-                    );
+                    let this = self.lock().unwrap();
```
Comment (hansl) [inline src/many-server/src/server.rs:332]: Let's meet about this. I don't think this is the right way of implementing this. There shouldn't be any changes to `server.rs` IMO.

File: src/many-server/src/server.rs [hunk 5]
```diff
@@ -276,40 +287,32 @@ impl LowLevelManyRequestHandler for Arc<Mutex<ManyServer>> {
...
+                }
+
+                Ok((
+                    cose_id.clone(),
+                    message,
+                    maybe_module,
+                    this.fallback.clone(),
+                ))
+            })()
+            .map_err(|many_err: ManyError| ResponseMessage::error(&cose_id.identity, id, many_err))
```
Comment (hansl) [inline src/many-server/src/server.rs:315]: We should move this into its own PR.

File: src/many-mock/tests/features/integration.feature [hunk 6]
```diff
@@ -0,0 +1,10 @@
+Feature: Mock Server
+
+Scenario: The server should answer with a JSON
+Given I request "object"
```
Comment (hansl) [inline src/many-mock/tests/features/integration.feature:4]: Just style: can we indent those?

**Review by hansl — CHANGES_REQUESTED** [review]
First round of comments.

**Review by hansl — APPROVED** [review]

**Pull request merged**